template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmHostGroupRoleConfigProviderProcessor.java

```diff
@@ -45,35 +65,40 @@ public void process(CmTemplateProcessor templateProcessor, TemplatePreparationOb
     }
 
-    private Map<String, Map<String, List<ApiClusterTemplateConfig>>> generateConfigs(CmTemplateProcessor templateProcessor, TemplatePreparationObject source) {
+    @VisibleForTesting
+    Map<String, Map<String, List<ApiClusterTemplateConfig>>> generateConfigs(CmTemplateProcessor templateProcessor, TemplatePreparationObject source) {
         Map<String, Map<String, List<ApiClusterTemplateConfig>>> configsByRoleConfigGroup = new HashMap<>();
-
-        Map<String, HostgroupView> hostGroups = source.getHostgroupViews().stream()
-                .collect(toMap(HostgroupView::getName, Function.identity()));
+        Map<String, HostgroupView> hostGroups = source.getHostgroupViews().stream().collect(toMap(HostgroupView::getName, Function.identity()));
         List<ApiClusterTemplateHostTemplate> hostTemplates = getHostTemplates(templateProcessor);
         Map<String, ServiceComponent> serviceComponents = templateProcessor.mapRoleRefsToServiceComponents();
 
         for (ApiClusterTemplateHostTemplate hostTemplate : hostTemplates) {
             String hostGroupName = hostTemplate.getRefName();
             List<String> roleConfigGroups = ofNullable(hostTemplate.getRoleConfigGroupsRefNames()).orElseGet(List::of);
             HostgroupView hostgroupView = hostGroups.get(hostGroupName);
+            groupByHostGroupName(source, configsByRoleConfigGroup, serviceComponents, hostGroupName, roleConfigGroups, hostgroupView);
+        }
+        return configsByRoleConfigGroup;
+    }
 
-            for (String roleConfigGroup : roleConfigGroups) {
-                for (CmHostGroupRoleConfigProvider provider : providers) {
-                    ServiceComponent serviceComponent = serviceComponents.get(roleConfigGroup);
-
-                    if (serviceComponent != null
-                            && Objects.equals(provider.getServiceType(), serviceComponent.getService())
-                            && provider.getRoleTypes().contains(serviceComponent.getComponent())) {
-
-                        configsByRoleConfigGroup.computeIfAbsent(roleConfigGroup, __ -> new HashMap<>())
-                                .computeIfAbsent(hostGroupName, __ -> new ArrayList<>())
-                                .addAll(provider.getRoleConfigs(serviceComponent.getComponent(), hostgroupView, source));
-                    }
+    private void groupByHostGroupName(TemplatePreparationObject source, Map<String, Map<String, List<ApiClusterTemplateConfig>>> configsByRoleConfigGroup,
+            Map<String, ServiceComponent> serviceComponents, String hostGroupName, List<String> roleConfigGroups, HostgroupView hostgroupView) {
+        for (String roleConfigGroup : roleConfigGroups) {
+            for (CmHostGroupRoleConfigProvider provider : providers) {
+                ServiceComponent serviceComponent = serviceComponents.get(roleConfigGroup);
+                if (serviceComponent != null
```

Comment:
Very large if condition. Can you please extract it?

Reply:
what is your proposal? I need to check all the three condition in the same if

Reply:
just extract this to a method an name it like areComponentsEqual or something like that

Reply:
done